roachtest: sqlsmith failed: interface conversion: coldata.Column is coldata.Int32s, not coldata.Int64s

roachtest.sqlsmith/setup=seed-multi-region/setting=multi-region [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3954616&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3954616&tab=artifacts#/sqlsmith/setup=seed-multi-region/setting=multi-region) on master @ [7bd8974eca566ca7625b6621decc5ae52d4fd38e](https://github.com/cockroachdb/cockroach/commits/7bd8974eca566ca7625b6621decc5ae52d4fd38e):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-multi-region/setting=multi-region/run_1
	sqlsmith.go:258,sqlsmith.go:297,test_runner.go:779: error: pq: internal error: unexpected error from the vectorized engine: interface conversion: coldata.Column is coldata.Int32s, not coldata.Int64s
		stmt:
		SELECT
			*
		FROM
			(
				SELECT
					tab_7367._timestamp AS col_13987,
					'00000000-0000-0000-0000-000000000004':::UUID::UUID AS col_13988,
					tab_7367._decimal AS col_13989,
					tab_7367._bool AS col_13990,
					lag(tab_7367._int2::INT4, tab_7367._int8::INT4, tab_7367._int4::INT4) OVER (PARTITION BY tab_7367._interval ORDER BY tab_7367._bytes, tab_7367._date RANGE UNBOUNDED PRECEDING)::INT8
						AS col_13991,
					tab_7367._int4 AS col_13992,
					nth_value(tab_7367._date::DATE, tab_7367._int2::INT4) OVER (PARTITION BY tab_7367._int8 ORDER BY tab_7367._bool DESC, tab_7367._bytes DESC)::DATE
						AS col_13993,
					tab_7367._string AS col_13994,
					2.0:::FLOAT8::FLOAT8 AS col_13995,
					tab_7367._int8 AS col_13996,
					2:::OID::OID AS col_13997,
					tab_7367._bytes AS col_13998,
					tab_7367._timestamptz AS col_13999,
					'00:00:06':::INTERVAL::INTERVAL AS col_14000
				FROM
					defaultdb.public.seed_mr_table@[0] AS tab_7367
			)
				AS tab_7368
		ORDER BY
			col_13987 NULLS FIRST,
			col_13988 NULLS FIRST,
			col_13989 NULLS FIRST,
			col_13990 NULLS FIRST,
			col_13991 NULLS FIRST,
			col_13992 NULLS FIRST,
			col_13993 NULLS FIRST,
			col_13994 NULLS FIRST,
			col_13995 NULLS FIRST,
			col_13996 NULLS FIRST,
			col_13997 NULLS FIRST,
			col_13998 NULLS FIRST,
			col_13999 NULLS FIRST,
			col_14000 NULLS FIRST;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed-multi-region/setting=multi-region.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -1274,6 +1274,11 @@ func NewColOperator(
 					PeersColIdx:     peersColIdx,
 				}
 
+				// Some window functions always return an INT result, so we use
+				// it as default and will update whenever necessary. We cannot
+				// use execinfrapb.GetWindowFunctionInfo because that function
+				// doesn't distinguish integers with different widths.
+				returnType := types.Int
 				if wf.Func.WindowFunc != nil {
 					// This is a 'pure' window function (e.g. not an aggregation).
 					windowFn := *wf.Func.WindowFunc
@@ -1296,6 +1301,7 @@ func NewColOperator(
 						)
 						result.Root, err = colexecwindow.NewRelativeRankOperator(
 							windowArgs, windowFn, wf.Ordering.Columns)
+						returnType = types.Float
 					case execinfrapb.WindowerSpec_NTILE:
 						opName := opNamePrefix + "ntile"
 						result.finishBufferedWindowerArgs(
@@ -1311,6 +1317,7 @@ func NewColOperator(
 						)
 						result.Root, err = colexecwindow.NewLagOperator(
 							windowArgs, argIdxs[0], argIdxs[1], argIdxs[2])
+						returnType = typs[argIdxs[0]]
 					case execinfrapb.WindowerSpec_LEAD:
 						opName := opNamePrefix + "lead"
 						result.finishBufferedWindowerArgs(
@@ -1319,6 +1326,7 @@ func NewColOperator(
 						)
 						result.Root, err = colexecwindow.NewLeadOperator(
 							windowArgs, argIdxs[0], argIdxs[1], argIdxs[2])
+						returnType = typs[argIdxs[0]]
 					case execinfrapb.WindowerSpec_FIRST_VALUE:
 						opName := opNamePrefix + "first_value"
 						result.finishBufferedWindowerArgs(
@@ -1327,6 +1335,7 @@ func NewColOperator(
 						)
 						result.Root, err = colexecwindow.NewFirstValueOperator(
 							windowArgs, wf.Frame, &wf.Ordering, argIdxs)
+						returnType = typs[argIdxs[0]]
 					case execinfrapb.WindowerSpec_LAST_VALUE:
 						opName := opNamePrefix + "last_value"
 						result.finishBufferedWindowerArgs(
@@ -1335,6 +1344,7 @@ func NewColOperator(
 						)
 						result.Root, err = colexecwindow.NewLastValueOperator(
 							windowArgs, wf.Frame, &wf.Ordering, argIdxs)
+						returnType = typs[argIdxs[0]]
 					case execinfrapb.WindowerSpec_NTH_VALUE:
 						opName := opNamePrefix + "nth_value"
 						result.finishBufferedWindowerArgs(
@@ -1343,6 +1353,7 @@ func NewColOperator(
 						)
 						result.Root, err = colexecwindow.NewNthValueOperator(
 							windowArgs, wf.Frame, &wf.Ordering, argIdxs)
+						returnType = typs[argIdxs[0]]
 					default:
 						return r, errors.AssertionFailedf("window function %s is not supported", wf.String())
 					}
@@ -1393,6 +1404,7 @@ func NewColOperator(
 						result.Root = colexecwindow.NewWindowAggregatorOperator(
 							windowArgs, aggType, wf.Frame, &wf.Ordering, argIdxs,
 							aggArgs.OutputTypes[0], aggFnsAlloc, toClose)
+						returnType = aggArgs.OutputTypes[0]
 					}
 				} else {
 					colexecerror.InternalError(errors.AssertionFailedf("window function spec is nil"))
@@ -1413,10 +1425,6 @@ func NewColOperator(
 					result.Root = colexecbase.NewSimpleProjectOp(result.Root, int(wf.OutputColIdx+tempColOffset), projection)
 				}
 
-				_, returnType, err := execinfrapb.GetWindowFunctionInfo(wf.Func, argTypes...)
-				if err != nil {
-					return r, err
-				}
 				result.ColumnTypes = appendOneType(result.ColumnTypes, returnType)
 				input = result.Root
 			}

--- a/pkg/sql/logictest/testdata/logic_test/window
+++ b/pkg/sql/logictest/testdata/logic_test/window
@@ -4301,3 +4301,34 @@ SELECT x, y, first_value(y) OVER (PARTITION BY x ROWS BETWEEN CURRENT ROW AND CU
 2  NotNull  NotNull
 2  NotNull  NotNull
 2  NotNull  NotNull
+
+# Regression test for incorrect type schema setup in the vectorized engine with
+# multiple window functions (#74087).
+statement ok
+CREATE TABLE t74087 AS
+  SELECT
+    g::INT2 AS _int2, g::INT4 AS _int4
+  FROM
+    ROWS FROM (generate_series(1, 5)) AS g;
+
+query IIIIIIIIIRRI rowsort
+SELECT
+  lag(_int2, _int4, _int2) OVER w,
+  lead(_int4, _int2, _int4) OVER w,
+  first_value(_int2) OVER w,
+  last_value(_int4) OVER w,
+  nth_value(_int2, _int4) OVER w,
+  min(_int4) OVER w,
+  row_number() OVER w,
+  rank() OVER w,
+  dense_rank() OVER w,
+  percent_rank() OVER w,
+  cume_dist() OVER w,
+  ntile(_int2) OVER w
+FROM t74087 WINDOW w AS (ORDER BY _int4);
+----
+1  2  1  1  1  1  1  1  1  0     0.2  1
+2  4  1  2  2  1  2  2  2  0.25  0.4  1
+3  3  1  3  3  1  3  3  3  0.5   0.6  1
+4  4  1  4  4  1  4  4  4  0.75  0.8  1
+5  5  1  5  5  1  5  5  5  1     1    1

